rust/nasl-builtin-cert/src/lib.rs

```diff
@@ -0,0 +1,422 @@
+    /// - fpr-sha-256 The SHA-256 fingerprint of the certificate.  The
+    ///                  fingerprint is, as usual, computed over the entire
+    ///                  DER encode certificate.
+    ///
+    /// - fpr-sha-1   The SHA-1 fingerprint of the certificate.  The
+    ///                  fingerprint is, as usual, computed over the entire
+    ///                  DER encode certificate.
+    ///
+    /// - image       Return the entire certificate as binary data.
+    ///
+    /// - algorithm-name  Same as signature-algorithm-name. TODO: Remove it and
+    ///                      leave only signature-algorithm-name.
+    ///
+    /// - signature-algorithm-name  Return the algorithm name used to sign the
+    ///                                certificate. Get the OID of the digest
+    ///                                algorithm and translated to a name from a
+    ///                                list from Wireshark.
+    ///                                See epan/dissectors/packet-pkcs1.c
+    ///
+    /// - public-key-algorithm-name  Return the algorithm name of the public key.
+    ///
+    /// - modulus      Return the RSA public key's modulus found in the
+    ///                   structure of the given cert.
+    ///
+    /// - exponent    Return the RSA public key's exponent found in
+    ///                  the structure of the given cert.
+    ///
+    /// - key-size    Return the size to hold the parameters size in bits.
+    ///                  For RSA the bits returned is the modulus.
+    ///                  For DSA the bits returned are of the public exponent.
+    ///
+    ///
+    /// The following arguments are required:
+    /// - pos(0): Object id of the certificate.
+    ///
+    /// - pos(1): A string with the command to select what to return; see above.
+    ///
+    /// The following arguments are optional:
+    /// - idx Used by certain commands to select the n-th value of a set
+    ///    of values.  If not given 0 is assumed.
+    ///
+    /// A NASL type depending on the used command.
+    #[nasl_function(named(idx))]
+    fn cert_query(
+        &self,
+        fd: usize,
+        query: &str,
+        idx: Option<usize>,
+    ) -> Result<NaslValue, FunctionErrorKind> {
+        let idx = idx.unwrap_or(0);
+        let handle = self.0.read().unwrap();
+        if fd >= handle.certs.len() {
+            return Err(FunctionErrorKind::WrongArgument(
+                "The given file descriptor is not valid.".to_string(),
+            ));
+        }
+        if let Some(cert) = &handle.certs[fd] {
+            match query {
+                "serial" => {
+                    let serial = cert.serial_number_asn1().clone().into_bytes();
+                    Ok(NaslValue::String(encode_hex(&serial)))
+                }
+                "subject" => {
+                    let der = cert.encode_der().unwrap();
+                    let (_, cert) = x509_parser::parse_x509_certificate(&der).unwrap();
```

Comment:
No, but we already parsed or decoded the `cert`, so it should always be possible to encode it back. The reason for the error is the internally used `write_encoded`, a function of the `Values` trait from the `bcder` library, which writes to any object implementing the `io::Write` trait. As the library internally writes to a Vec<u8> I thought, that this might be infallible, but actually I am not 100% sure.

Reply:
Oh right, read that wrong. Got it. I think your assumption is correct. Here is the `io::Write` impl for `Vec`, which is infallible.

```rust
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
        self.extend_from_slice(buf);
        Ok(buf.len())
    }
```

Reply:
I will add a comment to make it clear :+1: